File: cpp/src/renderer.cpp
#include "camintrinsics/renderer.hpp"

#include <opencv2/imgproc.hpp>

#include <cmath>

namespace ci {

cv::Point3d Pose::apply(const cv::Point3d& p) const {
  const cv::Vec3d v = R * cv::Vec3d(p.x, p.y, p.z) + t;
  return {v[0], v[1], v[2]};
}

cv::Point3d Pose::center() const {
  const cv::Vec3d c = -(R.t() * t);
  return {c[0], c[1], c[2]};
}

Pose lookAt(const cv::Point3d& eye, const cv::Point3d& target,
            const cv::Point3d& up) {
  cv::Vec3d z(target.x - eye.x, target.y - eye.y, target.z - eye.z);
  z /= cv::norm(z);
  const cv::Vec3d upv(-up.x, -up.y, -up.z);  // 'up' points -Y in our world
  cv::Vec3d x = upv.cross(z);
  if (cv::norm(x) < 1e-9) x = cv::Vec3d(1, 0, 0).cross(z);  // degenerate
  x /= cv::norm(x);
  const cv::Vec3d y = z.cross(x);
  Pose pose;
  pose.R = cv::Matx33d(x[0], x[1], x[2], y[0], y[1], y[2], z[0], z[1], z[2]);
  pose.t = -(pose.R * cv::Vec3d(eye.x, eye.y, eye.z));
  return pose;
}

Pose orbitPose(const cv::Point3d& target, double distance, double yawDeg,
               double pitchDeg) {
  const double yaw = yawDeg * CV_PI / 180.0, pitch = pitchDeg * CV_PI / 180.0;
  const cv::Point3d eye(target.x + distance * std::cos(pitch) * std::sin(yaw),
                        target.y - distance * std::sin(pitch),
                        target.z - distance * std::cos(pitch) * std::cos(yaw));
  return lookAt(eye, target);
}

namespace {

/// Split a camera-frame polyline into runs in front of the near plane.
/// Skip this and a point just behind the camera projects to a wild coordinate,
/// producing the long streaks that plague hand-written projective renderers.
std::vector<std::vector<cv::Point3d>> clipNear(
    const std::vector<cv::Point3d>& pts, double nearZ) {
  std::vector<std::vector<cv::Point3d>> runs;
  std::vector<cv::Point3d> cur;
  for (size_t i = 0; i < pts.size(); ++i) {
    if (pts[i].z >= nearZ) {
      cur.push_back(pts[i]);
    } else {
      if (!cur.empty()) {
        const cv::Point3d& a = cur.back();
        const double s = (nearZ - a.z) / (pts[i].z - a.z);
        cur.push_back(a + (pts[i] - a) * s);
        runs.push_back(cur);
        cur.clear();
      }
      if (i + 1 < pts.size() && pts[i + 1].z >= nearZ) {
        const double s = (nearZ - pts[i].z) / (pts[i + 1].z - pts[i].z);
        cur.push_back(pts[i] + (pts[i + 1] - pts[i]) * s);
      }
    }
  }
  if (!cur.empty()) runs.push_back(cur);
  runs.erase(std::remove_if(runs.begin(), runs.end(),
                            [](const std::vector<cv::Point3d>& r) {
                              return r.size() < 2;
                            }),
             runs.end());
  return runs;
}

}  // namespace

void render(cv::Mat& img, const std::vector<Polyline>& polylines,
            const Pose& pose, const Mat33& K, const Dist& D, double nearZ,
            double clipPixels) {
  for (const Polyline& pl : polylines) {
    std::vector<cv::Point3d> cam;
    cam.reserve(pl.pts.size() + 1);
    for (const cv::Point3d& p : pl.pts) cam.push_back(pose.apply(p));
    if (pl.closed && cam.size() > 2) cam.push_back(cam.front());

    for (const auto& run : clipNear(cam, nearZ)) {
      const std::vector<cv::Point2d> uv = projectPoints(run, K, D, nullptr);
      std::vector<cv::Point> poly;
      poly.reserve(uv.size());
      bool ok = true;
      for (const cv::Point2d& p : uv) {
        if (!std::isfinite(p.x) || !std::isfinite(p.y) ||
            std::abs(p.x) > clipPixels || std::abs(p.y) > clipPixels) {
          ok = false;
          break;
        }
        poly.emplace_back(cvRound(p.x), cvRound(p.y));
      }
      if (!ok || poly.size() < 2) continue;
      if (pl.filled && poly.size() >= 3)
        cv::fillPoly(img, std::vector<std::vector<cv::Point>>{poly}, pl.color,
                     cv::LINE_AA);
      else
        cv::polylines(img, poly, false, pl.color, pl.thickness, cv::LINE_AA);
    }
  }
}

void drawCrosshair(cv::Mat& img, const Mat33& K, int size) {
  const KParams p = splitK(K);
  cv::drawMarker(img, {img.cols / 2, img.rows / 2}, colors::kDim,
                 cv::MARKER_CROSS, size, 1);
  const cv::Point pp(cvRound(p.cx), cvRound(p.cy));
  cv::drawMarker(img, pp, colors::kRed, cv::MARKER_CROSS, size + 6, 2);
  cv::circle(img, pp, size, colors::kRed, 1, cv::LINE_AA);
}

void drawTextBlock(cv::Mat& img, const std::vector<std::string>& lines,
                   cv::Point org, double scale, const cv::Scalar& color,
                   int lineH, bool bg) {
  if (lines.empty()) return;
  if (bg) {
    int wmax = 0, baseline = 0;
    for (const auto& t : lines)
      wmax = std::max(wmax, cv::getTextSize(t, cv::FONT_HERSHEY_SIMPLEX, scale,
                                            1, &baseline).width);
    cv::Mat overlay = img.clone();
    cv::rectangle(overlay, cv::Point(org.x - 6, org.y - 16),
                  cv::Point(org.x + wmax + 8,
                            org.y + lineH * int(lines.size()) - 6),
                  cv::Scalar(18, 18, 18), -1);
    cv::addWeighted(overlay, 0.62, img, 0.38, 0, img);
  }
  for (size_t i = 0; i < lines.size(); ++i)
    cv::putText(img, lines[i], {org.x, org.y + int(i) * lineH},
                cv::FONT_HERSHEY_SIMPLEX, scale, color, 1, cv::LINE_AA);
}

cv::Mat hstackLabeled(const std::vector<cv::Mat>& images,
                      const std::vector<std::string>& labels, int pad,
                      int labelH) {
  const cv::Scalar bg(24, 24, 28);
  int hs = 0, total = 0;
  for (const auto& im : images) hs = std::max(hs, im.rows);
  for (const auto& im : images) total += im.cols;
  total += pad * (int(images.size()) - 1);

  cv::Mat out(hs + labelH, total, CV_8UC3, bg);
  int x = 0;
  for (size_t i = 0; i < images.size(); ++i) {
    images[i].copyTo(out(cv::Rect(x, labelH, images[i].cols, images[i].rows)));
    cv::putText(out, i < labels.size() ? labels[i] : "", {x + 8, 18},
                cv::FONT_HERSHEY_SIMPLEX, 0.5, colors::kWhite, 1, cv::LINE_AA);
    x += images[i].cols + pad;
  }
  return out;
}

cv::Mat vstack(const std::vector<cv::Mat>& images, const cv::Scalar& bg) {
  int w = 0, h = 0;
  for (const auto& im : images) {
    w = std::max(w, im.cols);
    h += im.rows;
  }
  cv::Mat out(h, w, CV_8UC3, bg);
  int y = 0;
  for (const auto& im : images) {
    im.copyTo(out(cv::Rect(0, y, im.cols, im.rows)));
    y += im.rows;
  }
  return out;
}

}  // namespace ci

File: python/camintrinsics/renderer.py
"""A ~150-line software renderer, written so that K and D are the only optics.

There is no OpenGL and no shading here.  Every pixel drawn goes through
:func:`camintrinsics.intrinsics.project_points`, so whatever you see on screen
is *literally* the intrinsic model doing its job.
"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass

import cv2
import numpy as np

from .intrinsics import project_points, split_K
from .scene import COLORS, Polyline

__all__ = ["Pose", "look_at", "orbit_pose", "render", "draw_crosshair",
           "draw_text_block", "hstack_labeled", "save_image"]


@dataclass
class Pose:
    """A rigid world -> camera transform: ``p_cam = R @ p_world + t``."""
    R: np.ndarray
    t: np.ndarray

    def apply(self, points_world: np.ndarray) -> np.ndarray:
        P = np.atleast_2d(np.asarray(points_world, dtype=np.float64))
        return P @ self.R.T + self.t.reshape(1, 3)

    @property
    def center(self) -> np.ndarray:
        """Camera centre in world coordinates: ``C = -R^T t``."""
        return -self.R.T @ self.t.reshape(3)


def look_at(eye, target, up=(0, -1, 0)) -> Pose:
    """Build a Pose that puts the camera at ``eye`` looking at ``target``.

    OpenCV convention: +Z forward (into the scene), +X right, +Y **down** --
    hence the default ``up=(0,-1,0)`` for a world whose +Y also points down.
    """
    eye, target, up = (np.asarray(v, dtype=np.float64) for v in (eye, target, up))
    z = target - eye
    z /= np.linalg.norm(z)
    x = np.cross(-up, z)          # 'up' points -Y, so -up is the world's +Y-ish
    nx = np.linalg.norm(x)
    if nx < 1e-9:                 # degenerate: looking straight along 'up'
        x = np.cross(np.array([1.0, 0.0, 0.0]), z)
        nx = np.linalg.norm(x)
    x /= nx
    y = np.cross(z, x)
    R = np.stack([x, y, z], axis=0)
    return Pose(R, -R @ eye)


def orbit_pose(target, distance: float, yaw_deg: float, pitch_deg: float) -> Pose:
    """Camera orbiting ``target`` -- the mouse-free way to fly around a scene."""
    yaw, pitch = np.radians(yaw_deg), np.radians(pitch_deg)
    target = np.asarray(target, dtype=np.float64)
    offset = np.array([distance * np.cos(pitch) * np.sin(yaw),
                       -distance * np.sin(pitch),
                       -distance * np.cos(pitch) * np.cos(yaw)])
    return look_at(target + offset, target)


# --------------------------------------------------------------------------
def _clip_near(pts_cam: np.ndarray, near: float) -> list[np.ndarray]:
    """Split a camera-frame polyline into runs in front of the near plane.

    Without this step, a point just behind the camera projects to a wild
    coordinate and you get long streaks shooting across the image -- the
    single most common bug in hand-written projective renderers.
    """
    z = pts_cam[:, 2]
    runs, cur = [], []
    for i in range(len(pts_cam)):
        if z[i] >= near:
            cur.append(pts_cam[i])
        else:
            if cur:
                # extend the run up to the near plane before cutting it
                a, b = cur[-1], pts_cam[i]
                s = (near - a[2]) / (b[2] - a[2])
                cur.append(a + s * (b - a))
                runs.append(np.array(cur))
                cur = []
        if z[i] < near and i + 1 < len(pts_cam) and z[i + 1] >= near:
            a, b = pts_cam[i], pts_cam[i + 1]
            s = (near - a[2]) / (b[2] - a[2])
            cur.append(a + s * (b - a))
    if cur:
        runs.append(np.array(cur))
    return [r for r in runs if len(r) >= 2]


def render(img: np.ndarray, polylines: list[Polyline], pose: Pose,
           K: np.ndarray, D=None, near: float = 0.05,
           clip_pixels: float = 6000.0) -> np.ndarray:
    """Draw a scene into ``img`` through the camera described by ``pose, K, D``."""
    h, w = img.shape[:2]
    for pl in polylines:
        pts_cam = pose.apply(pl.points)
        if pl.closed and len(pts_cam) > 2:
            pts_cam = np.vstack([pts_cam, pts_cam[:1]])
        for run in _clip_near(pts_cam, near):
            uv = project_points(run, K, D)
            if not np.all(np.isfinite(uv)):
                continue
            if np.abs(uv).max() > clip_pixels:
                continue
            poly = np.round(uv).astype(np.int32)
            if pl.fill is not None and len(poly) >= 3:
                cv2.fillPoly(img, [poly], pl.fill, lineType=cv2.LINE_AA)
            else:
                cv2.polylines(img, [poly], False, pl.color, pl.thickness,
                              lineType=cv2.LINE_AA)
    return img


# --------------------------------------------------------------------------
# Small drawing helpers shared by the apps
# --------------------------------------------------------------------------
def draw_crosshair(img: np.ndarray, K: np.ndarray, size: int = 14) -> None:
    """Mark the principal point (cx, cy) and the geometric image centre.

    Seeing these two markers drift apart is the fastest way to understand that
    ``(cx, cy)`` is *not* "the middle of the image" -- it is where the optical
    axis happens to pierce the sensor.
    """
    h, w = img.shape[:2]
    fx, fy, cx, cy, _ = split_K(K)
    gx, gy = int(round(w / 2)), int(round(h / 2))
    cv2.drawMarker(img, (gx, gy), COLORS["dim"], cv2.MARKER_CROSS, size, 1)
    px, py = int(round(cx)), int(round(cy))
    cv2.drawMarker(img, (px, py), COLORS["red"], cv2.MARKER_CROSS, size + 6, 2)
    cv2.circle(img, (px, py), size, COLORS["red"], 1, cv2.LINE_AA)


def draw_text_block(img: np.ndarray, lines, org=(10, 22), scale: float = 0.45,
                    color=COLORS["white"], line_h: int = 17,
                    bg: bool = True) -> None:
    """Draw a left-aligned multi-line HUD with a readable dark backdrop."""
    x, y = org
    if bg and lines:
        wmax = max(cv2.getTextSize(t, cv2.FONT_HERSHEY_SIMPLEX, scale, 1)[0][0]
                   for t in lines)
        overlay = img.copy()
        cv2.rectangle(overlay, (x - 6, y - 16),
                      (x + wmax + 8, y + line_h * len(lines) - 6), (18, 18, 18), -1)
        cv2.addWeighted(overlay, 0.62, img, 0.38, 0, img)
    for i, text in enumerate(lines):
        cv2.putText(img, text, (x, y + i * line_h), cv2.FONT_HERSHEY_SIMPLEX,
                    scale, color, 1, cv2.LINE_AA)


def hstack_labeled(images, labels, pad: int = 8, label_h: int = 26,
                   bg=(24, 24, 24)) -> np.ndarray:
    """Stack images side by side, each with a caption bar above it."""
    hs = max(im.shape[0] for im in images)
    tiles = []
    for im, lab in zip(images, labels):
        tile = np.full((hs + label_h, im.shape[1], 3), bg, np.uint8)
        tile[label_h:label_h + im.shape[0], :im.shape[1]] = im
        cv2.putText(tile, lab, (8, 18), cv2.FONT_HERSHEY_SIMPLEX, 0.5,
                    COLORS["white"], 1, cv2.LINE_AA)
        tiles.append(tile)
    out = np.full((hs + label_h, sum(t.shape[1] for t in tiles) + pad * (len(tiles) - 1), 3),
                  bg, np.uint8)
    x = 0
    for t in tiles:
        out[:, x:x + t.shape[1]] = t
        x += t.shape[1] + pad
    return out


def save_image(path: str, img: np.ndarray) -> bool:
    """Write an image, reporting honestly.

    ``cv2.imwrite`` returns False for a path it cannot write and raises for an
    extension it does not recognise.  Either way the caller must not go on to
    print "wrote ..." -- a tool that lies about what it did is worse than one
    that fails.
    """
    directory = os.path.dirname(os.path.abspath(path))
    try:
        os.makedirs(directory, exist_ok=True)
        if cv2.imwrite(path, img):
            print(f"wrote {path}  ({img.shape[1]}x{img.shape[0]})")
            return True
        print(f"could not write {path} -- is the directory writable?",
              file=sys.stderr)
    except (OSError, cv2.error) as exc:
        print(f"could not write {path} -- {exc}", file=sys.stderr)
    return False
